File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;
import java.util.*;

public class Main {
    static int answer = 0;
    static int N;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());

        DFS(0, new int[N]);

        System.out.println(answer);
    }

    private static void DFS(int L, int[] arr) {
        if(L == arr.length) {
            answer++;
            return;
        }

        for (int i = 0; i < N; i++) {
            arr[L] = i;
            if(isPromise(L, arr)) {
                DFS(L + 1, arr);
            }
        }
    }

    private static boolean isPromise(int L, int[] arr) {
        for (int i = 0; i < L; i++) {
            if(arr[i] == arr[L]) return false;
            else if(Math.abs(i - L) == Math.abs(arr[i] - arr[L])) return false;
        }

        return true;
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.py

N = int(input())

rows = [0] * N
answer = 0

def is_promissing(row: int, col: int) -> bool:
    global rows

    for r in range(row):
        if r == row or rows[r] == col:
            return False
        elif abs(r - row) == abs(rows[r] - col):
            return False

    return True


def dfs(L: int):
    global rows, N, answer

    if L == N:
        answer += 1
        return

    for i in range(N):
        rows[L] = i
        if is_promissing(L, i):
            dfs(L + 1)

dfs(0)
print(answer)